[DB-26-19] Warn when projection state becomes greater than 8mb

lakshdeepsingheventstore opened the pull request:
Added : A warning to the log when a projection state size becomes greater than 8 MB. https://eventstore.aha.io/develop/requirements/DB-26-19

timothycoleman commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -20,21 +23,31 @@ public class ResultEventEmitter : IResultEventEmitter {
+							_namesBuilder.EffectiveProjectionName, _namesBuilder.EffectiveProjectionName,
+							System.Text.ASCIIEncoding.Unicode.GetByteCount(result.Event.Data) / Math.Pow(10, 6));
+					}
+					warningProjectionList.Add(_namesBuilder.EffectiveProjectionName);
```
isn't this going to add the projection to the list even if the state is small, and then when the state is big it won't log anything?

lakshdeepsingheventstore replied:
Yeah it was a mistake on my end, rectified.

timothycoleman commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -20,21 +23,31 @@ public class ResultEventEmitter : IResultEventEmitter {
 							: new EmittedDataEvent(
 								streamId, Guid.NewGuid(), "Result", true, projectionResult, null, at, null),
 						_resultStreamMetadata);
+				if (!string.IsNullOrEmpty(result.Event.Data) && !warningProjectionList.Contains(_namesBuilder.EffectiveProjectionName)) {
```
this will do a linear search through the list every time a projection emits

lakshdeepsingheventstore replied:
I have pushed new changes as per which the list would only be searched when an event of event type "Result" is emitted, and not for any other types of emitted events.

timothycoleman reviewed (changes requested):
- in the pr description, this isn't a bug so we don't need to put fixed. we can use `Added:`
- im not sure, but does this apply to projection state specifically or all emitted projection events?

lakshdeepsingheventstore commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -20,21 +23,31 @@ public class ResultEventEmitter : IResultEventEmitter {
 			CheckpointTag at) {
 			var streamId = _namesBuilder.MakePartitionResultStreamName(partition);
 			var allResultsStreamId = _namesBuilder.GetResultStreamName();
+			EmittedEventEnvelope result;
```
Reverted back to previous format. It was required as per some of the changes which I had done locally to test this issue, but later was not needed.

lakshdeepsingheventstore commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -43,6 +56,16 @@ public class ResultEventEmitter : IResultEventEmitter {
+					}
+					warningProjectionList.Add(_namesBuilder.EffectiveProjectionName);
+				}
+
```
Done

lakshdeepsingheventstore commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -20,21 +23,31 @@ public class ResultEventEmitter : IResultEventEmitter {
 								streamId, Guid.NewGuid(), "Result", true, projectionResult, null, at, null),
 						_resultStreamMetadata);
+				if (!string.IsNullOrEmpty(result.Event.Data) && !warningProjectionList.Contains(_namesBuilder.EffectiveProjectionName)) {
+					if (System.Text.ASCIIEncoding.Unicode.GetByteCount(result.Event.Data) / Math.Pow(10, 6) >= 8) {
```
Done

lakshdeepsingheventstore commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -1,8 +1,11 @@
 namespace EventStore.Projections.Core.Services.Processing {
 	public class ResultEventEmitter : IResultEventEmitter {
 		private readonly ProjectionNamesBuilder _namesBuilder;
+		private static List<string> warningProjectionList = new List<string>();
```
It could be a possible case. Changed the variable to a non static one.

timothycoleman commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -1,8 +1,11 @@
 namespace EventStore.Projections.Core.Services.Processing {
 	public class ResultEventEmitter : IResultEventEmitter {
 		private readonly ProjectionNamesBuilder _namesBuilder;
+		private List<string> WarningProjectionList = new List<string>();
```
for non-static member vars it should be _warningProjectionList

is it possible for multiple different names to be added to the list?

lakshdeepsingheventstore replied:
Based on the investigation and tests so far, it looks like each projection has a different ResultEventEmitter object associated with it. So in the latest changes, I have got ride of the List variable and instead introduced a bool variable to the ResultEventEmitter Class to keep track if the projection's state size is less than 8 MB

timothycoleman reviewed (changes requested):
^

timothycoleman commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -29,6 +41,9 @@ public class ResultEventEmitter : IResultEventEmitter {
 							: new EmittedDataEvent(
 								streamId, Guid.NewGuid(), "Result", true, projectionResult, null, at, null),
 						_resultStreamMetadata);
+				if (!string.IsNullOrEmpty(result.Event.Data) && result.Event.EventType.Equals("Result")) {
```
i think this is the same check as below isn't it

lakshdeepsingheventstore replied:
Yeah it is indeed the same check. The purpose of keeping this check outside of the method CheckStateSize() was to avoid calling the CheckStateSize() method for every emitted event, but rather only call it for the events that determine the size of the projection state (that is events of type "Result").

timothycoleman commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -16,6 +18,16 @@ public class ResultEventEmitter : IResultEventEmitter {
 		}
 
+		private void CheckStateSize(EmittedEventEnvelope result) {
+			if (System.Text.ASCIIEncoding.Unicode.GetByteCount(result.Event.Data) >= 8_000_000 && !_largeStateWarningLogged) {
```
arguments are evaluated from left to right, so it would make sense to check the bool first since it's quicker. counting the bytes is quite expensive. it could even be an early return

timothycoleman commented on src/EventStore.Projections.Core/Services/Processing/ResultEventEmitter.cs:
```diff
@@ -1,8 +1,10 @@
 namespace EventStore.Projections.Core.Services.Processing {
 	public class ResultEventEmitter : IResultEventEmitter {
 		private readonly ProjectionNamesBuilder _namesBuilder;
+		private bool _largeStateWarningLogged = false; //This would be set to true as soon as the projection state size becomes greater than 8MB and a warning is logged, preventing flooding of logs with warnings each time an event is emitted.
```
i think the comment isn't necessary for this now that the variable name is clear

hayley-jean reviewed (changes requested):
Thanks! There's one more place where the state size needs to be checked

What you have here works for the projection result that will be written to the `$projection-{projection_name}-result` stream. This result is only written for Queries, or for Continuous/OneTime projections that call `outputState()` in the projection definition.

The size of the state that makes up the checkpoints written to the `$projection-{projection_name}-checkpoint` stream and `$projection-{projection_name}-{partition_name}-checkpoint` streams also needs to be checked.

Probably the best place to check and log a warning for the state size is when `StateUpdated` is called in the [CoreProjectionCheckpointManager](https://github.com/EventStore/EventStore/blob/8d83ef069dfb1a7ce028f46bb18ce0d5169e0585/src/EventStore.Projections.Core/Services/Processing/CoreProjectionCheckpointManager.cs#L161) (in addition to the check you're doing here in `ResultEventEmitter`)

lakshdeepsingheventstore commented:
>The size of the state that makes up the checkpoints written to the $projection-{projection_name}-checkpoint stream and $projection-{projection_name}-{partition_name}-checkpoint streams also needs to be checked.
Making changes to [CoreProjectionCheckpointManager](https://github.com/EventStore/EventStore/blob/8d83ef069dfb1a7ce028f46bb18ce0d5169e0585/src/EventStore.Projections.Core/Services/Processing/CoreProjectionCheckpointManager.cs#L161) solves the same purpose as was being sone by the changes in `ResultEventEmitter` file. So in the latest commit/changes, I have removed the changes from `ResultEventEmitter`, and added those in `CoreProjectionCheckpointManager` in the method `StateUpdated`
However, to exclusively check for state sizes for the checkpoint streams, separate checks had to be added to `CoreProjectionCheckpointWriter` and `PartitionStateUpdateManager` files, since the check in `CoreProjectionCheckpointManager` does not log anything when the projection Checkpoint is written.

lakshdeepsingheventstore commented on src/EventStore.Projections.Core/Services/Processing/PartitionStateUpdateManager.cs:
```diff
@@ -29,6 +33,16 @@ private class State {
 
+		private void CheckPartitionStateSize(PartitionState state) {
+			if (!_largePartitionStateWarningLogged && state.State.Length >= 8_000_000) {
+				Log.Warning(
```
Done

hayley-jean commented on src/EventStore.Projections.Core/Services/Processing/CoreProjectionCheckpointManager.cs:
```diff
@@ -175,6 +177,17 @@ public abstract class CoreProjectionCheckpointManager : IProjectionCheckpointMan
+			CheckStateSize(_currentProjectionState);
+		}
+
+		private void CheckStateSize(PartitionState result) {
```
You could actually do the partition check here as well instead of in `PartitionStateUpdateManager`. If you move the `if (partition == "" && newState.State == null)` condition, you could check the state size of `newState` before the other two conditions. e.g:

```
if (partition == "" && newState.State == null) // ignore non-root partitions and non-changed states
	throw new NotSupportedException("Internal check");

CheckStateSize(newState, partition);

if (_usePersistentCheckpoints && partition != "")
	CapturePartitionStateUpdated(partition, oldState, newState);

if (partition == "")
	_currentProjectionState = newState;
```

lakshdeepsingheventstore replied:
Done

hayley-jean reviewed (changes requested):
Sorry about not including this in my earlier review, but I learned new things about how the `result` of a projection interacts with the `state` while testing this. Basically, we need to include both the `state` and `result` in `PartitionState` when checking the size of the projection state.

I previously thought that the result was only written if you had `outputState()` defined on your projection, but it turns out that the [check for that is commented out](https://github.com/EventStore/EventStore/blob/28fa63d6989d8c675cff4f606540394fe6166bec/src/EventStore.Projections.Core/Services/Processing/EventProcessingProjectionProcessingPhase.cs#L260).

Instead, a result is always written for continuous projections (just not always in a result stream). If the projection includes a `transformBy()` function then the result will be the result of that transformation run on the state. Otherwise it [simply duplicates the state into the result](https://github.com/EventStore/EventStore/blob/28fa63d6989d8c675cff4f606540394fe6166bec/src/EventStore.Projections.Core/Services/Processing/EventProcessingProjectionProcessingPhase.cs#L264).

The result is [included with the state](https://github.com/EventStore/EventStore/blob/28fa63d6989d8c675cff4f606540394fe6166bec/src/EventStore.Projections.Core/Services/Processing/PartitionState.cs#L77) when the checkpoint is written, so this can effectively double the state size that is written to disk.

I think the best way to handle this would be to have a `Size` property on `PartitionState` that returns something like `_state.Length + _result?.Length ?? 0;` and use that for the warning log. Otherwise, by the time the warning is logged at 8mb, you are already potentially over the 16mb limit.

Let me know if you want a sample projection to test this case
